File: Projects/TicTacToe.js
/*
 * Tic Tac Toe:
 *
 * Your project, should you choose to code it, is implement Tic Tac Toe.
 * The game should be playable by two people on the command line.
 * Run the program by opening
 */


// Some magic to get a 'prompt' method, that gives you input
var prompt = require('prompt-sync')();

// The 3x3 game board of Tic Tac Toe.
// Value of 0 is no move made
//          1 is Player 1 'O'
//         -1 is Player 2 'X'
var board = [[0, 0, 0], 
			 [0, 0, 0], 
			 [0, 0, 0]];

/* Printed board should look like this: 
 O | X | 
 ---------
   | O |  
 ---------
   |   | Os
*/

// Tracks the current player (1 for 'Player 1', -1 for 'Player 2')
var player = 1;
var numMoves = 0;

function printBoard(board) {
	/* Write code here to print the gameboard to screen using
	   the console.log() command.
	   Returns nothing
	 */

	for(var i = 0; i < board.length; i++) {
		var row = " ";
		for(var j = 0; j < board.length; j++) {	
			if(j !== 0) {
				row += "|";		
			}
			if(board[i][j] === 1) {
				row += " O ";
			}
			else if(board[i][j] === -1) {
				row += " X ";
			}
			else {
				row += "   ";
			}
		}
			if(i !== 0) {
				console.log("-------------");
			}
			console.log(row);
	}
}

function checkCols(board) {
	/* Write code to check if either player has a column win
	   Return 1 or -1 for the player who won 
	 */
	 for(var j = 0; j < board.length; j++) {
	 	var sum = 0;
	 	for(var i = 0; i < board.length; i++) {
	 		sum += board[i][j];
	 	}
	 	// console.log("Completed column "+ i + " " + board[i] + " sum = " + sum)
 		if(sum === board.length || sum === -(board.length)) {
 			return (sum / board.length);
 		}

	 }
	return 0; 
}

function checkRows(board) {
	/* Write code to check if either player has a row win
	   Return 1 or -1 for the player who won 
	 */
	 for(var i = 0; i < board.length; i++) {
	 	var sum = 0;
	 	for(var j = 0; j < board.length; j++) {
	 		sum += board[i][j];
	 	}
	 	// console.log("Completed column "+ i + " " + board[i] + " sum = " + sum)
 		if(sum === board.length || sum === -(board.length)) {
 			return (sum / board.length);
 		}

	 }
	return 0;
}

function checkDiags(board) {
	/* Write code to check if either player has a diag win
	   Return 1 or -1 for the player who won 
	 */
	var sum1 = 0;
	var sum2 = 0;
	for(var i = 0; i < board.length; i++) {
	 	sum1 += board[i][i];
	 	sum2 += board[i][board.length - i - 1];		
	}
	if(sum1 === board.length || sum1 === -(board.length)) {
 		return (sum1 / board.length);
 	}
 	else if(sum2 === board.length || sum2 === -(board.length)) {
 		return (sum2 / board.length);
 	}
 	return 0;

}

function checkWin(board) {
	/* Write code here to call checkCols, checkRows, checkDiags
	   and see if either player has won.
	   Return 1 or -1 for the player who won
	*/
	var win1 = checkCols(board);
	if(win1 !== 0) {
		return checkCols(board);
	}
	var win2 = checkRows(board);
	if(win2 !== 0) {
		return checkRows(board);
	}
	var win3 = checkDiags(board);
	if(win3 !== 0) {
		return checkDiags(board);
	}
	return 0;
}

function ifCanWinMove(board) {
	for(var i = 0; i < board.length; i++) {
		for(var j = 0; j < board.length; j++) {
			if(board[i][j] === 0) {
				board[i][j] = player;
				if(checkWin(board) == player) {
					return true;
				}
				else {
					board[i][j] = 0;
				}
			}
		}
	}
}

function ifCanBlockMove(board) {
	//console.log("AI cannot make winning move, attempting block as " + player)
	for(var i = 0; i < board.length; i++) {
		for(var j = 0; j < board.length; j++) {
			if(board[i][j] === 0) {
				board[i][j] = player * -1;
				if(checkWin(board) == player * -1) {
					board[i][j] = player;
					return true;
				}
				else {
					board[i][j] = 0;
				}
			}
		}
	}
}

function randomAIMove(board) {
	//console.log("AI making random move as " + player)
	for(var i = 0; i < board.length; i++) {
		for(var j = 0; j < board.length; j++) {
			if(board[i][j] === 0) {
				board[i][j] = player;
				return true;
			}
		}
	}
}

function aiMove(board) {
	//console.log("AI making move as player " + player)
	numMoves++;
	if(ifCanWinMove(board)) {
		//console.log("AI makng winning move as " + player)
		return board;
	}
	else if(ifCanBlockMove(board)) {
		//console.log("AI making blocking move as " + player)
		return board;
	}
	else {
		randomAIMove(board);
		return board;
	}
}

function makeMove(board) {
	/* Write code here, to ask the player for a move,
	   then attempt to place their move on the gameboard
	   Return the gameboard
	 */
	var needInput = true;
	var x;
	var y;
	while(needInput) {
		console.log("Please make a move, Player " + (player == 1 ? 1 : 2) + ". (e.g; a1, b2, c3...)");
		var input = prompt();
		switch(input[0]) {
		 	case "a":
		 		x = 0;
		 		break;
		 	case "b":
		 		x = 1;
		 		break;
		 	case "c":
		 		x = 2;
		 		break;	 		
		 }
		y = input[1] - 1;
		
		if(x >= 0 && x < board.length && y >= 0 && y < board.length && board[x][y] === 0) {
			needInput = false;
			board[x][y] = player;
			numMoves++;
		}		
	}
	return board;
}

function startGame() {
	/* Make the game run in this function
	 * You can delete my code from below:
	 */
	 var answer = prompt("Hello, do you want to play a game of TicTacToe? [yes/no]");
	 if(answer === "yes") {
	 	console.log("Great! Let's start!");
		while(true) {
			printBoard(board);
			if(player == 1) {
				makeMove(board);
			}
			else {
				aiMove(board);				
			}
			
			console.log(checkWin(board));
			if(checkWin(board) !== 0) {
				printBoard(board);
				console.log("Player " + (checkWin(board) == 1 ? 1 : 2) + " has won!");
				return false; 
			}
			
			if(numMoves=== board.length * board.length) {
				printBoard(board);
				console.log("It's a draw!");
				return false;
			}
			player *= -1;
		}

	 }
	else if(answer === "no") {
		console.log("Aww...");
	}
	else {
		console.log("Sorry, I didn't understand that...");
	}

}

startGame();

File: Misc/TicTacToe.py
class TicTacToe:
	def __init__(self):
		global board
		self.board = [[0 for j in range(3)] for i in range(3)]
		global length
		self.length = len(self.board)
		global player
		self.player = 1
		global numMoves
		self.numMoves = 0

	def printBoard(self):
		print('')
		for i in range(self.length):
			row = []
			for j in self.board[i]:
				if j == 0:
					row.append(' ')
				elif j == 1:
					row.append('O')
				elif j == -1:
					row.append('X')
			if i != 0:
				print(' ---------')
			print(' ' + ' | '.join(row))
		print('')


	def checkColumns(self):
		for j in range(self.length):
			sumColumn = 0
			for i in range(self.length):
				sumColumn += self.board[i][j]
			if sumColumn == self.length or sumColumn == -self.length:
				return sumColumn // self.length

		return 0

	def checkRows(self):
		for r in self.board:
			sumRow = sum(r)
			if sumRow == self.length or r == -self.length:
				return sumRow // self.length

		return 0

	def checkDiagonals(self):
		sumDiag1 = 0
		sumDiag2 = 0
		for i in range(self.length):
			sumDiag1 += self.board[i][i]
			sumDiag2 += self.board[i][self.length - i - 1]
		if sumDiag1 == self.length or sumDiag1 == -self.length:
			return sumDiag1 // self.length
		if sumDiag2 == self.length or sumDiag2 == -self.length:
			return sumDiag2 // self.length

		return 0

	def checkWin(self):
		checks = [self.checkColumns(), self.checkRows(), self.checkDiagonals()]
		for i in checks:
			if i == 1 or i == -1:
				return i

		return 0

	def aiMove(self):
		self.numMoves += 1
		if self.canWin():
			return self.board
		elif self.canBlock():
			return self.board
		else:
			self.randomMove()
			return self.board

	def canWin(self):
		for i in range(self.length):
			for j in range(self.length):
				if self.board[i][j] == 0:
					self.board[i][j] = self.player
					if self.checkWin() == self.player:
						return True
					else:
						self.board[i][j] = 0

	def canBlock(self):
		for i in range(self.length):
			for j in range(self.length):
				if self.board[i][j] == 0:
					self.board[i][j] = self.player * -1
					if self.checkWin() == self.player * -1:
						self.board[i][j] = self.player
						return True
					else:
						self.board[i][j] = 0

	def randomMove(self):
		for i in range(self.length):
			for j in range(self.length):
				if self.board[i][j] == 0:
					self.board[i][j] = self.player
					return

	def makeMove(self):
		x = -1
		y = -1
		while True:
			move = input(' Player ' + str(1 if self.player == 1 else 2) + ', please enter your move (eg a1 => a is row, 1 is column): ')
			move = move.strip()
			if len(move) != 2:
				print(' Sorry, not sure what your input is. Please try again.')
				continue
			if move[0] == 'a':
				x = 0
			elif move[0] == 'b':
				x = 1
			elif move[0] == 'c':
				x = 2
			else:
				print(' Sorry, not sure what your input is. Please try again.')
				continue
			y = int(move[1]) - 1
			if y < 0 or y > 2:
				print(' Sorry, not sure what your input is. Please try again.')
				continue
			if self.board[x][y] == 0:
				self.numMoves += 1
				self.board[x][y] = self.player
				break


	def startGame(self):
		answer = input(' Hello, do you want to play a game of TicTacToe? yes/no: ')
		if answer == 'yes':
			answer2 = input(" Great! Do you want to play against the computer? yes/no: ")
			print(" Let's start!")
			if answer2 == 'yes':
				ai = True
			else:
				ai = False
			while True:
				self.printBoard()
				if ai:
					if self.player == 1:
						self.makeMove()
					else:
						self.aiMove()

				else:
					self.makeMove()

				if self.checkWin() != 0:
					self.printBoard()
					print(' Player ' + str(1 if self.checkWin() == 1 else 2) + ' has won!')
					break

				if self.numMoves == self.length ** 2:
					self.printBoard()
					print(" It's a draw!")
					break

				self.player *= -1

		elif answer == 'no':
			print(' Aww... :(')
		else:
			print(" Sorry, I didn't understand that...")

if __name__ == '__main__':
	t = TicTacToe()
	t.startGame() 